project/dexter-vs/Dexter/Common/Core/Utils/DexterHierarchyService.cs

```diff
@@ -0,0 +1,150 @@
+            return (string)projectPath;
+        }
+
+        private IEnumerable<string> TraverseSourceFiles(IVsHierarchy pHierarchy, uint itemid)
+        {
+            IntPtr nestedHierarchyObj;
+            uint nestedItemId;
+            var hierGuid = typeof(IVsHierarchy).GUID;
+
+            var hr = pHierarchy.GetNestedHierarchy(itemid, ref hierGuid, out nestedHierarchyObj, out nestedItemId);
+            if (VSConstants.S_OK == hr && IntPtr.Zero != nestedHierarchyObj)
+            {
+                var nestedHierarchy = Marshal.GetObjectForIUnknown(nestedHierarchyObj) as IVsHierarchy;
+                Marshal.Release(nestedHierarchyObj);
+                if (nestedHierarchy != null)
+                {
+                    foreach (var filePath in TraverseSourceFiles(nestedHierarchy, nestedItemId))
+                    {
+                        yield return filePath;
+                    }
+                }
+            }
+            else
+            {
+                var currentFilePath = GetFilePath(pHierarchy, itemid);
+                if (IsSourceFilePath(currentFilePath))
+                    yield return currentFilePath;
+
+                object pVar;
+                hr = pHierarchy.GetProperty(itemid,
+                                           (int)__VSHPROPID.VSHPROPID_FirstVisibleChild,
+                                           out pVar);
+                ErrorHandler.ThrowOnFailure(hr);
+                if (VSConstants.S_OK == hr)
+                {
+                    var childId = GetItemId(pVar);
+                    while (childId != VSConstants.VSITEMID_NIL)
+                    {
+                        foreach (var filePath in TraverseSourceFiles(pHierarchy, childId))
+                        {
+                            yield return filePath;
+                        }
+
+                        hr = pHierarchy.GetProperty(childId,
+                                                   (int)__VSHPROPID.VSHPROPID_NextVisibleSibling,
+                                                   out pVar);
+                        if (VSConstants.S_OK == hr)
+                        {
+                            childId = GetItemId(pVar);
+                        }
+                        else
+                        {
+                            ErrorHandler.ThrowOnFailure(hr);
+                            break;
+                        }
+                    }
+                }
+            }
+        }
+
+        private bool IsSourceFilePath(string filePath)
+        {
+            if (filePath != null && filePath.EndsWith(".cs"))
+                return true;
+
+            return false;
+        }
+
+        private string GetFilePath(IVsHierarchy pHierarchy, uint itemid)
+        {
+            object filePath;
+            pHierarchy.GetProperty(itemid, (int)__VSHPROPID.VSHPROPID_SaveName, out filePath);
+            return (string)filePath;
+        }
+
+        private static uint GetItemId(object pvar)
```

Comment:
This method comes from the reference code. I do not know exactly what the IVsHierarchy's internal implementation is, so it would be better to use this method.